src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500RelativeDistinguishedName.cs

```diff
@@ -0,0 +1,153 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Formats.Asn1;
+
+namespace System.Security.Cryptography.X509Certificates
+{
+    /// <summary>
+    ///   Represents a Relative Distinguished Name component of an X.500 Distinguished Name.
+    /// </summary>
+    /// <seealso cref="X500DistinguishedName"/>
+    public sealed class X500RelativeDistinguishedName
+    {
+        private Oid? _singleElementType;
+        private readonly ReadOnlyMemory<byte> _singleElementValue;
+
+        /// <summary>
+        ///   Gets the encoded representation of this Relative Distinguished Name.
+        /// </summary>
+        /// <value>
+        ///   The encoded representation of this Relative Distinguished Name.
+        /// </value>
+        public ReadOnlyMemory<byte> RawData { get; }
+
+        internal X500RelativeDistinguishedName(ReadOnlyMemory<byte> rawData)
+        {
+            RawData = rawData;
+
+            ReadOnlySpan<byte> rawDataSpan = rawData.Span;
+            AsnValueReader outer = new AsnValueReader(rawDataSpan, AsnEncodingRules.DER);
+
+            // Windows does not enforce the sort order on multi-value RDNs.
+            AsnValueReader rdn = outer.ReadSetOf(skipSortOrderValidation: true);
+            AsnValueReader typeAndValue = rdn.ReadSequence();
+
+            Oid firstType = Oids.GetSharedOrNewOid(ref typeAndValue);
+            ReadOnlySpan<byte> firstValue = typeAndValue.ReadEncodedValue();
+            typeAndValue.ThrowIfNotEmpty();
+
+            if (rdn.HasData)
+            {
+                while (rdn.HasData)
+                {
+                    typeAndValue = rdn.ReadSequence();
+
+                    if (Oids.GetSharedOrNullOid(ref typeAndValue) is null)
```

Comment:
This took me a minute to understand what is going on. This is just avoiding an allocation while trying to spin through the RDN while consuming the OID, correct?

Maybe this is worth a comment. Alternatively, I guess you could just peek the tag, check it, and consume the value.